src/checkout.c

```diff
@@ -1411,36 +1410,83 @@ static int blob_content_to_file(
 	struct stat *st,
 	git_blob *blob,
 	const char *path,
-	const char * hint_path,
+	const char *hint_path,
 	mode_t entry_filemode)
 {
+	int flags = data->opts.file_open_flags;
 	mode_t file_mode = data->opts.file_mode ?
 		data->opts.file_mode : entry_filemode;
-	git_buf out = GIT_BUF_INIT;
+	struct checkout_stream writer;
+	mode_t mode;
 	git_filter_list *fl = NULL;
+	int fd;
 	int error = 0;
 
 	if (hint_path == NULL)
 		hint_path = path;
 
-	if (!data->opts.disable_filters)
-		error = git_filter_list__load_with_attr_session(
+	if ((error = mkpath2file(data, path, data->opts.dir_mode)) < 0)
+		return error;
+
+	if (flags <= 0)
+		flags = O_CREAT | O_TRUNC | O_WRONLY;
+	if (!(mode = file_mode))
+		mode = GIT_FILEMODE_BLOB;
+
+	if ((fd = p_open(path, flags, mode)) < 0) {
+		giterr_set(GITERR_OS, "Could not open '%s' for writing", path);
+		return fd;
+	}
+
+	if (!data->opts.disable_filters &&
+		(error = git_filter_list__load_with_attr_session(
 			&fl, data->repo, &data->attr_session, blob, hint_path,
-			GIT_FILTER_TO_WORKTREE, GIT_FILTER_OPT_DEFAULT);
+			GIT_FILTER_TO_WORKTREE, GIT_FILTER_OPT_DEFAULT)))
+		return error;
+
+	if (fl)
+		git_filter_list__set_temp_buf(fl, &data->tmp);
+
+	/* setup the writer */
+	memset(&writer, 0, sizeof(struct checkout_stream));
+	writer.base.write = checkout_stream_write;
+	writer.base.close = checkout_stream_close;
+	writer.base.free = checkout_stream_free;
+	writer.path = path;
+	writer.fd = fd;
+	writer.open = 1;
+
+	error = git_filter_list_stream_blob(fl, blob, (git_filter_stream *)&writer);
 
-	if (!error)
-		error = git_filter_list_apply_to_blob(&out, fl, blob);
+	assert(writer.open == 0);
 
 	git_filter_list_free(fl);
+	p_close(fd);
```

Comment:
Why perform the close here instead of giving ownership of the descriptor to the filter and have its `->close()` close the descriptor.

Here you're also throwing away the return value from close, which can return an error (though I'm sure we do this in very many other places).